docs/source/conf.py

```diff
@@ -5,7 +5,9 @@
 # https://www.sphinx-doc.org/en/master/usage/configuration.html
 import datetime
 import emoji
-
+import sys
+import os
+sys.path.insert(0, os.path.abspath(os.path.join("..", "..")))
```

Comment:
```suggestion
from pathlib import Path
import sys

sys.path.insert(0, Path(__file__).parent.parent.absolute())
```

Gets rid of the old-style `os` module.
I presume we need to add the src directory to the path though so the Sphinx addon can parse the commands.
Could we not add `../../data-safe-haven` instead then?

Reply:
Since `docs` is in a separate directory, can we not `pip install .` as part of setting up this `hatch` environment and then just `import data_safe_haven`?

Reply:
Possibly. I'm guessing the important thing is that `data_safe_haven` is in the Python path.

Not sure how you would represent that in the dependencies.

Speaking of which, how has the lockfile been updated when the docs env in `pyproject.toml` doesn't have all of those dependencies?

Reply:
Looks like `hatch` automatically installs the package into its environment unless you specify `detached = True`.

```
> hatch -e default shell
> pip freeze
...
cryptography==43.0.3
-e git+ssh://git@github.com/jemrobinson/data-safe-haven.git@6586af9affce845574a799a07e6ea03691b9c298#egg=data_safe_haven
debugpy==1.8.7
...
```

Reply:
Yes, I removed `detached = True` from `[tool.hatch.envs.docs]` in `pyproject.toml` previously. `import data_safe_haven` doesn't work for me, locally - it still complains the module is not found; perhaps I'm missing something.

Reply:
@craddm, so if you do the following it doesn't work?

```
> hatch -e docs shell
> python
>>> import data_safe_haven
```

Reply:
no, that works, but the docs won't build

Reply:
What about things like `import data_safe_haven.commands.sre` like in the directives?

Reply:
> What about things like `import data_safe_haven.commands.sre` like in the directives?

Which directives exactly? the Typer ones don't permit any content

Reply:
OK, let's take a step back.

You're saying that there's an error that can only be solved by manually adding the current directory to the path (which lets you do `from data_safe_haven import X`)

However, it also looks like you're already able to run `from data_safe_haven import X` in the `docs` environment.

What's the actual error message that you get if you remove the `sys.path.insert` line?

Reply:
```
/workspaces/data-safe-haven/docs/source/reference/config.md:5: CRITICAL: Failed to import "data_safe_haven.commands.config:config_command_group"The following exception was raised:
Traceback (most recent call last):
  File "/home/deploydsh/.local/share/hatch/env/pip-compile/data-safe-haven/Ce8xTVjM/docs/lib/python3.12/site-packages/sphinxcontrib/typer/__init__.py", line 309, in import_object
    obj = import_module(try_path)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'data_safe_haven' [docutils]
```

Reply:
I mean these directives,

```markdown
:::{typer} data_safe_haven.commands.sre:sre_command_group
:width: 65
:prog: dsh sre
:show-nested:
:make-sections:
:::
```

It looks a bit like a Python import so I'm wondering whether `import data_safe_haven` is fine but `import data_safe_haven.commands.sre` isn't for some reason.
If not, it might be doing something more low level that relies on looking for files in the Python path.

Reply:
> I mean these directives,
> 
> ```
> :::{typer} data_safe_haven.commands.sre:sre_command_group
> :width: 65
> :prog: dsh sre
> :show-nested:
> :make-sections:
> :::
> ```
> 
> It looks a bit like a Python import so I'm wondering whether `import data_safe_haven` is fine but `import data_safe_haven.commands.sre` isn't for some reason. If not, it might be doing something more low level that relies on looking for files in the Python path.

Yes, those directives don't allow any content. I'm not sure where I would put `import data_safe_haven`

Reply:
Does `/home/deploydsh/.local/share/hatch/env/pip-compile/data-safe-haven/Ce8xTVjM/docs/lib/python3.12/site-packages/` have a package called `data_safe_haven` in it?

Reply:
```
/workspaces/data-safe-haven/docs/source/reference/config.md:5: CRITICAL: Failed to import "data_safe_haven:config_command_group"The following exception was raised:
Traceback (most recent call last):
  File "/home/deploydsh/.local/share/hatch/env/pip-compile/data-safe-haven/Ce8xTVjM/docs/lib/python3.12/site-packages/sphinxcontrib/typer/__init__.py", line 309, in import_object
    obj = import_module(try_path)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'data_safe_haven' [docutils]
```

So, no.

Reply:
> Does `/home/deploydsh/.local/share/hatch/env/pip-compile/data-safe-haven/Ce8xTVjM/docs/lib/python3.12/site-packages/` have a package called `data_safe_haven` in it?

No

Reply:
OK, I just checked out this branch and did the following:

1. Remove the `sys.path.insert` line

2. Run `hatch run docs:build`

This worked as expected.

If I look at the `docs` environment I see

```
> ls -1 /Users/jrobinson/Library/Application\ Support/hatch/env/pip-compile/data-safe-haven/sPksTL1q/docs/lib/python3.12/site-packages/
Arpeggio-2.0.2.dist-info
CHANGELOG.rst
CONTRIBUTING.md
...
_data_safe_haven.pth
...
data_safe_haven-5.1.0.dist-info
...
```

Reply:
> > Does `/home/deploydsh/.local/share/hatch/env/pip-compile/data-safe-haven/Ce8xTVjM/docs/lib/python3.12/site-packages/` have a package called `data_safe_haven` in it?
> 
> No

That is strange, would suggest the `detached=False` isn't working or the env is out of date (although hatch should check for that each time you run a command or script :shrug:).

With what @jemrobinson said it might just mean recreating the env `hatch env remove docs && hatch run docs:build`.

Reply:
Looks like it doesn't update installed versions of the module after creating the environment (and apparently doesn't install them if you switch from `detached=True` to `detached=False`) https://github.com/pypa/hatch/issues/771

Reply:
So yes, it works now I've cleared and recreated the environment!